File: design-pattern/proxy-pattern/protection-proxy/protection.ts
abstract class DocumentReader {
    abstract unlockPdf(): void
};

class RealDocumentReader extends DocumentReader {
    public unlockPdf(): void {
        console.log("you are reading this pdf")
    }
};

class User {
    public name: string;
    public isSubscribed: boolean;

    constructor(name: string, subscribed: boolean) {
        this.name = name;
        this.isSubscribed = subscribed;
    }
};

class ProtectionProxy extends DocumentReader {
    private documentReader: RealDocumentReader;
    private user: User;

    constructor(user: User) {
        super();
        this.user = user;
        this.documentReader = new RealDocumentReader();
    };

    unlockPdf(): void {
        if (this.user.isSubscribed) {
            this.documentReader.unlockPdf();
        } else {
            console.log("you are authorized to access this pdf");
        }
    }
};

const user = new User("omkar", true);
const proxy = new ProtectionProxy(user);
proxy.unlockPdf();

File: design-pattern/proxy-pattern/protection-proxy/protection.cpp
#include <iostream>
using namespace std;

class IDocumentReader
{
public:
    virtual void unlock_document() = 0;
    virtual ~IDocumentReader() {}
};

class RealDocument : public IDocumentReader
{
public:
    void unlock_document() override
    {
        cout << "You are reading the document" << endl;
    }
};

class User
{
public:
    string name;
    bool premiumMembership;

    User(string name, bool membership)
    {
        this->name = name;
        this->premiumMembership = membership;
    }
};

class ProxyDocument : public IDocumentReader
{
private:
    User *user;
    RealDocument realDoc;

public:
    ProxyDocument(User *user)
    {
        this->user = user;
    }

    void unlock_document() override
    {
        if (user->premiumMembership)
        {
            realDoc.unlock_document();
        }
        else
        {
            cout << "Access denied: Premium membership required" << endl;
        }
    }
};

int main()
{
    User *user = new User("Omkar", true);

    IDocumentReader *doc = new ProxyDocument(user);
    doc->unlock_document();

    delete doc;
    delete user;
}
